fe/fe-core/src/main/java/com/starrocks/connector/paimon/PaimonMetadata.java

```diff
@@ -292,4 +301,94 @@ public long getTableCreateTime(String dbName, String tblName) throws Exception {
         }
         return 0;
     }
+
+    public List<PartitionInfo> getChangedPartitionInfo(Table table, long mvSnapShotID) {
+        LOG.debug("Get changed partitionInfo start, table:{}, mvLatestSnapShotID {}", table, mvSnapShotID);
+        List<PartitionInfo> result = new ArrayList<>();
+        PaimonTable paimonTable = (PaimonTable) table;
+        Map<String, Long> partitionToSnapshotId = fetchChangedPartitionWithVersion(paimonTable, mvSnapShotID);
+        for (Map.Entry<String, Long> entry : partitionToSnapshotId.entrySet()) {
+            PaimonPartitionInfo partitionInfo = new PaimonPartitionInfo(entry.getKey(), entry.getValue());
+            result.add(partitionInfo);
+            if (entry.getValue() != null) {
+                mvSnapShotID = Math.max(mvSnapShotID, entry.getValue());
+            }
+        }
+        LOG.debug("Get changed partitionInfo:{}", result);
+        return result;
+    }
+
+    private Map<String, Long> fetchChangedPartitionWithVersion(PaimonTable paimonTable, long mvSnapshotId) {
+        Map<String, Long> partitionToSnapshotId = new HashMap<>();
+        FileMonitorTable fileMonitorTable = new FileMonitorTable(paimonTable.getNativeTable());
+        Long latestId = fileMonitorTable.snapshotManager().latestSnapshotId();
+        long latestSnapshotId = latestId == null ? Long.MIN_VALUE : latestId;
+        LOG.debug("Paimon table {} latest snapshotId {}, currentId {}",
+                paimonTable.getName(), latestSnapshotId, mvSnapshotId);
+        if (mvSnapshotId >= latestSnapshotId && latestSnapshotId != Long.MIN_VALUE) {
+            LOG.info("Paimon table {} currentId {} > latest snapshotId {} ",
+                    paimonTable.getName(), mvSnapshotId, latestSnapshotId);
+            return partitionToSnapshotId;
+        }
+        ReadBuilder readBuilder = fileMonitorTable.newReadBuilder();
+        StreamTableScan scan = readBuilder.newStreamScan();
+        TableRead read = readBuilder.newRead();
+        if (mvSnapshotId != Long.MIN_VALUE) {
+            scan.restore(mvSnapshotId + 1);
+        } else {
+            scan.restore(null);
+        }
+        try {
+            // It may cost too many time to scan rows if paimon snapshot too frequently or mv has long refresh interval.
+            while (true) {
+                if (!scanMonitorTable(paimonTable, scan, read, partitionToSnapshotId)) {
+                    break;
+                }
+            }
+        } catch (RuntimeException e) {
+            // Query rewrite depends on this function to check if mv is stale, throw exception may kill normal query.
```

Comment:
The MV prepare procedure would catch all exception, and give up the mv if prepare failed